<b>Exploratory Data Analysis</b>

I performed initial EDA in the data wrangling portion of this project and now I will take a deeper dive to discover more patterns, spot anomalies and to check assumptions with the help of summary statistics and graphical representations. I will try to answer the questions below.

<ul>
    <li>What are the top 10 highest rated book titles? Top 10 worst rated book titles?</li>
    <li>Which book titles had the most ratings count? Which had the least?</li>
    <li>What is the least amount of ratings a user has done? Has each user rated at least 1 book?</li>  
    <li>What is the most amount of ratings a user has done?</li>  
    <li>What is the distribution of rating counts for users?</li>
    <li>Which authors had the highest average ratings? Which had the lowest?</li>
    <li>Which authors had the highest ratings count? Which had the lowest?</li>
    <li>Which authors wrote the most books? Which wrote the least?</li>
</ul>


In [2]:
# import required libraries
import pandas as pd
import numpy
import plotly
import cufflinks as cf
from plotly.offline import download_plotlyjs,init_notebook_mode,plot,iplot
init_notebook_mode(connected=True)
cf.go_offline()

In [3]:
# read in the csv file
df = pd.read_csv('Data/bookratings_clean.csv', low_memory=False)
df.head()

user_id  book_id  rating                          authors  \
0        1      258       5  Carlos Ruiz Zafón, Lucia Graves   
1       11      258       3  Carlos Ruiz Zafón, Lucia Graves   
2      143      258       4  Carlos Ruiz Zafón, Lucia Graves   
3      242      258       5  Carlos Ruiz Zafón, Lucia Graves   
4      325      258       4  Carlos Ruiz Zafón, Lucia Graves   

                                               title  average_rating  \
0  The Shadow of the Wind (The Cemetery of Forgot...            4.24   
1  The Shadow of the Wind (The Cemetery of Forgot...            4.24   
2  The Shadow of the Wind (The Cemetery of Forgot...            4.24   
3  The Shadow of the Wind (The Cemetery of Forgot...            4.24   
4  The Shadow of the Wind (The Cemetery of Forgot...            4.24   

   ratings_count language_code  num_of_authors           author_0  ...  \
0         263685           eng               2  Carlos Ruiz Zafón  ...   
1         263685           eng               2  Carlos Ruiz Zafón  ...   
2         263685           eng               2  Carlos Ruiz Zafón  ...   
3         263685           eng               2  Carlos Ruiz Zafón  ...   
4         263685           eng               2  Carlos Ruiz Zafón  ...   

  author_37 author_38 author_39 author_40 author_41 author_42 author_43  \
0        -1        -1        -1        -1        -1        -1        -1   
1        -1        -1        -1        -1        -1        -1        -1   
2        -1        -1        -1        -1        -1        -1        -1   
3        -1        -1        -1        -1        -1        -1        -1   
4        -1        -1        -1        -1        -1        -1        -1   

  author_44 author_45 author_46  
0        -1        -1        -1  
1        -1        -1        -1  
2        -1        -1        -1  
3        -1        -1        -1  
4        -1        -1        -1  

[5 rows x 56 columns]

<b>What are the top 10 highest rated book titles?</b>

In [4]:
# get columns title and average rating, drop duplicates
df_title_avgrating = df[['title', 'average_rating']]
df_title_avgrating = df_title_avgrating.drop_duplicates()

In [5]:
# list top 10 average rated book titles
df_title_avgrating.nlargest(10, ['average_rating']).set_index('title')['average_rating']

title
The Complete Calvin and Hobbes                                       4.82
Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5)               4.77
Words of Radiance (The Stormlight Archive, #2)                       4.77
Mark of the Lion Trilogy                                             4.76
ESV Study Bible                                                      4.76
It's a Magical World: A Calvin and Hobbes Collection                 4.75
There's Treasure Everywhere: A Calvin and Hobbes Collection          4.74
Harry Potter Boxset (Harry Potter, #1-7)                             4.74
Harry Potter Collection (Harry Potter, #1-6)                         4.73
The Authoritative Calvin and Hobbes: A Calvin and Hobbes Treasury    4.73
Name: average_rating, dtype: float64

<b>What are the top 10 worst rated book titles?</b>

In [6]:
# list top 10 worst average rated book titles
df_title_avgrating.nsmallest(10, ['average_rating']).set_index('title')['average_rating']

title
One Night at the Call Center                                          2.47
The Almost Moon                                                       2.67
The Finkler Question                                                  2.76
Lost                                                                  2.80
Revenge Wears Prada: The Devil Returns (The Devil Wears Prada, #2)    2.84
Among the Ten Thousand Things                                         2.84
The Red House                                                         2.93
Pygmy                                                                 2.96
The 3 Mistakes of My Life                                             2.97
Beautiful You                                                         2.98
Name: average_rating, dtype: float64

<b>Which book titles had the most ratings count?</b>

In [7]:
# get columns title and ratings count, drop duplicates
df_title_ratingscount = df[['title', 'ratings_count']]
df_title_ratingscount = df_title_ratingscount.drop_duplicates()

In [8]:
# get top 10 highest rating counts for book titles
df_plot = df_title_ratingscount.nlargest(10, ['ratings_count']).set_index('title')['ratings_count']

# plot top 10 
df_plot.iplot(
    kind='bar',
    bins=5,
    xTitle='book title',
    yTitle='count',
    color='green',
    linecolor='black',
    title='Top 10 Books for Highest Ratings Count')

<b>Which book titles had the lowest ratings count?</b>

In [9]:
# get top 10 highest rating counts for book titles
df_plot = df_title_ratingscount.nsmallest(10, ['ratings_count']).set_index('title')['ratings_count']

# plot top 10 
df_plot.iplot(
    kind='bar',
    bins=5,
    xTitle='book title',
    yTitle='count',
    color='green',
    linecolor='black',
    title='Top 10 Books for Lowest Ratings Count')

<b>What is the least amount of ratings a user has done? Has every user rated at least one book?</b>  

In [10]:
# get rating counts for users
df_user_ratingcounts = df.groupby('user_id').size().to_frame()

# set size column name to ratings_count
df_user_ratingcounts.columns = ['ratings_count'] 

# reset index 
df_user_ratingcounts = df_user_ratingcounts.reset_index()
df_user_ratingcounts.head()

user_id  ratings_count
0        1            114
1        2             56
2        3             85
3        4            126
4        5             91

In [11]:
df_user_ratingcounts['ratings_count'].sort_values(ascending=True)[:10]

25855    1
47142    1
49879    2
34710    2
47974    2
28714    2
32127    2
52082    2
42735    2
39282    2
Name: ratings_count, dtype: int64

It tooks like each user has rated at least 1 book. Specifically, there are only 2 users who have rated only 1 book. 

<b>What is the most amount of ratings a user has done? </b>

In [12]:
df_user_ratingcounts['ratings_count'].sort_values(ascending=False)[:10]

12380    195
52035    194
12873    194
30943    193
53291    189
10508    189
45553    189
33064    188
28157    188
7562     188
Name: ratings_count, dtype: int64

It looks like one user rated 195 different books, which was the most ratings any user has done.

<b>What is the distribution of rating counts for users?</b>

In [13]:
# plot the distribution
df_user_ratingcounts.iplot(
    x='user_id',
    y='ratings_count',
    xTitle='users',
    yTitle='ratings count',
    title='Distribution of User Ratings Count',
    color='green',
    mode='markers',
)

The distribution of rating counts for users looks evenly distributed between 0 and 200. 

<b>Which authors had the highest average ratings?</b>

Since most of the books had up to 3 authors, I will only do calculations for the first 3.

In [15]:
# get columns title and average rating, drop duplicates
df_authors_avgrating = df[['author_0', 'author_1', 'author_2', 'average_rating']]
df_authors_avgrating = df_authors_avgrating.drop_duplicates()

I will restructure the dataframe so that all authors will be in one column and average rating in another column. To do this, I will first separate the the authors into separate dataframes and then concatenate them in the end. Then I will calculate the average rating of average rating for each author.

In [16]:
# get subset of first author column and average rating 
df_author_0 = df_authors_avgrating[['author_0', 'average_rating']]

# rename author_0 column to author
df_author_0.rename(columns={'author_0':'author'}, inplace=True)

Do the same as above for author_1 and author_2.

In [17]:
# subset of author_1
df_author_1 = df_authors_avgrating[['author_1', 'average_rating']]
df_author_1.rename(columns={'author_1':'author'}, inplace=True)

In [18]:
# subset of author_2
df_author_2 = df_authors_avgrating[['author_2', 'average_rating']]
df_author_2.rename(columns={'author_2':'author'}, inplace=True)

In [19]:
# concatenate all author dataframes 
df_authors_avgrating = pd.concat([df_author_0, df_author_1, df_author_2], ignore_index=True, sort=False)

Earlier I flagged books that did not have author_1 or author 2 values to -1. I will remove those records from the dataframe.

In [20]:
df_authors_avgrating = df_authors_avgrating[df_authors_avgrating['author'] != '-1']

I will now calculate the average rating of average rating for each author, and display the top average rated authors.

In [21]:
# create a groupby variable that groups average rating values by authors
groupby_author_rating = df_authors_avgrating['average_rating'].groupby(df_authors_avgrating['author'])

# get the average of average rating by author
author_avgrating = groupby_author_rating.mean().sort_values(ascending=False)[:10]
print(author_avgrating)

author
Lane T. Dennis     4.76
Wayne A. Grudem    4.76
Bill Watterson     4.70
Ronald A. Beers    4.67
Hafez              4.63
Angie Thomas       4.62
G.B. Trudeau       4.61
Alisa Kwitney      4.61
Richard Ferrone    4.60
Richard Case       4.59
Name: average_rating, dtype: float64


<b>Which authors had the lowest average ratings?</b>

In [22]:
# get the average of average rating by author
author_avgrating = groupby_author_rating.mean().sort_values(ascending=True)[:10]
print(author_avgrating)

author
Howard Jacobson      2.760000
Julia Pierpont       2.840000
Anne Enright         3.040000
Wednesday Martin     3.050000
Chetan Bhagat        3.086667
Jennifer Close       3.120000
Adolf Hitler         3.130000
Yann Martel          3.130000
Chuck Wendig         3.150000
Christos Tsiolkas    3.160000
Name: average_rating, dtype: float64


<b>Which authors had the highest ratings count?</b>

In [23]:
# get columns title and average rating, drop duplicates
df_authors_rc = df[['author_0', 'author_1', 'author_2', 'ratings_count']]

In [24]:
# get subset of first author column and ratings count
df_author_0 = df_authors_rc[['author_0', 'ratings_count']]
df_author_0.rename(columns={'author_0':'author'}, inplace=True)

# subset of author_1
df_author_1 = df_authors_rc[['author_1', 'ratings_count']]
df_author_1.rename(columns={'author_1':'author'}, inplace=True)

# subset of author_2
df_author_2 = df_authors_rc[['author_2', 'ratings_count']]
df_author_2.rename(columns={'author_2':'author'}, inplace=True)

In [25]:
# concatenate all author dataframes 
df_authors_rc = pd.concat([df_author_0, df_author_1, df_author_2], ignore_index=True, sort=False)

In [26]:
df_authors_rc = df_authors_rc[df_authors_rc['author'] != '-1']

In [27]:
# create a groupby variable that groups ratings count by authors
groupby_author_rc = df_authors_rc['ratings_count'].groupby(df_authors_rc['author'])

# get the top 10 highest rating counts by author
author_rc = groupby_author_rc.sum().sort_values(ascending=False)[:10]
print(author_rc)

author
J.K. Rowling           267625574685
Mary GrandPré          263255257814
Suzanne Collins        166996701466
Stephenie Meyer        102823956800
J.R.R. Tolkien          63175393790
Harper Lee              61365879925
George Orwell           55898310335
Dan Brown               52146953974
F. Scott Fitzgerald     44692634300
Jane Austen             38017950236
Name: ratings_count, dtype: int64


<b>Which authors had the lowest ratings count?</b>

In [29]:
# get the top 10 lowest rating counts by author
author_rc = groupby_author_rc.sum().sort_values(ascending=True)[:10]
print(author_rc)

author
Hafez                207975
Donna Russo Morin    297000
Stacy King           321168
SunNeko Lee          321168
Crystal S. Chan      321168
Jana Deleon          324948
Mary Jordan          349508
Gina DeJesus         349508
Amanda Berry         349508
Hikaru Suruga        370312
Name: ratings_count, dtype: int64
